### DISCUSSION TITLE
Constructing Images fromarray/buffer in Greyscale vs. Palettized

### DISCUSSION THREAD

**Discussion by TheRealQuantam** [python-pillow/Pillow#6561] [Q&A] [answered]
I'm noticing in Pillow 9.2.0 that fromarray et al with P mode create a palette and attach it to the image, which does not happen when the buffer is loaded as L. This makes P much slower than L to load even when the default palette is immediately overwritten by putpalette (and it's thus highly recommended to load as L and then use putpalette). If this hasn't already been changed since then I'd suggest either

1. Defer palette initialization until the palette is actually used with P
2. Allow a palette to be passed to fromarray et al
3. Note in the documentation that it's recommended to load the image as L if you're going to putpalette (at least I didn't see anything like that)

**Comment by radarhere** [comment]
Could we get a code example to demonstrate the speed difference?

> I'm noticing in Pillow 9.2.0

Let us know if you mean that this speed difference just started in 9.2.0. I suspect you don't mean that.

**Reply by TheRealQuantam** [reply]
[PillowPvsL.zip](https://github.com/python-pillow/Pillow/files/9534947/PillowPvsL.zip)

```
Mode L create time 0.001255 secs, put time 0.004581 secs
Mode P create time 0.024284 secs, put time 0.004640 secs
```
This is in CPython with Python 3.9.7. And I meant that Pillow 9.2.0 is what I have installed and the only thing I tried it on.

**Reply by radarhere** [reply]
For whatever reason, the zip link isn't working.

**Reply by TheRealQuantam** [reply]
Hmm. GitHub was on the fritz that night. I just reuploaded it.

**Reply by radarhere** [reply]
I've created PR #6567. Please run the speed test with that and see what you think.

**Reply by TheRealQuantam** [reply]
I'm... not sure how to do that, actually. My method of learning Python has always been "I'll learn how to do it when I need to do it", and I haven't needed to test an experimental change to a pypi package, yet :P

**Reply by TheRealQuantam** [reply]
Okay, I think I figured it out.
```
Mode L create time 0.001288 secs, put time 0.000886 secs
Mode P create time 0.002539 secs, put time 0.000863 secs
```
Almost an order of magnitude speedup. I'd still mention that L is significantly faster than P in the documentation, though.

**Reply by radarhere** [reply]
Cool. I was going to suggest that if you have wheel, pip and the dependencies installed, `python3 -m pip install git+https://github.com/radarhere/Pillow.git@palette` is simplest, but you figured something out.

Is that PR sufficiently fast then?

**Reply by TheRealQuantam** [reply]
Almost an order of magnitude speedup. I'd still mention that L is significantly faster than P (twice as fast) in the documentation, though.

**Reply by radarhere** [reply]
On your machine, with the PR, this is a 1ms speed difference when loading 139 images. On my machine, it is an 0.5ms speed difference. I think documenting it would make users think it is a better problem than that.

**Reply by radarhere** [reply, marked as answer]
#6567 has now been merged.
